### DISCUSSION TITLE
Apple ARKit in Avalonia.iOS through NativeControlHost

### DISCUSSION THREAD

**Discussion by llfab** [AvaloniaUI/Avalonia#17043] [General]
Just wanted to take the time to mention that I was able to use Apple ARKit in an Avalonia iOS app. At this moment that is really only proof of concept:
![image](https://github.com/user-attachments/assets/71cde348-6272-4514-b906-792b62c8c966)

This is the priciple:
- Create an xplat app with the newest templates
- Remove all projects except the lib with the views and view models, the iOS project AND the desktop project (the latter seems necessary for the previewer to work in Rider)
- I used the principle of NativeControlHost working with a custom interface to create the specific instance in the iOS project (not in the lib project). The lib project does not target iOS hence, ARKit is not available there. If you make the lib project also target iOS, the previewer will not work (at least I have not found a way to make the previewer work without a Desktop project)
- Create the ARSCNView with content in the iOS project and use this as the interface implementation for the NativeControlHost

At this point I cannot really say what the limitations of ARKit in this environment are. Once we progress, I will add more input here.

A simple sample project can be found here: https://github.com/llfab/Samples/tree/main/AvaloniaiOSApplication1

**Comment by maxkatz6** [comment]
> I used the principle of NativeControlHost working with a custom interface to create the specific instance in the iOS project (not in the lib project). The lib project does not target iOS hence, ARKit is not available there

You also can add "net8.0-ios" target to the lib project, making it multitarget.
It will require some "#if IOS" conditions to access ARKit there, but generally a working solution.

> At this point I cannot really say what the limitations of ARKit in this environment are. Once we progress, I will add more input here.

No Avalonia content can be displayed on top of native controls. That's probably the single big limitation for some apps.

**Reply by llfab** [reply]
Thanks Max, thats helpful. If you multi-target the lib, is there anything needed in the Desktop vs. iOS assembly to make sure it uses the right target? 

No overlay controls: yes, that is clear as this has always been the case even in WPF, right. Refresh rate will be interesting...

**Reply by maxkatz6** [reply]
No, nothing special is needed.
"net8.0-ios" mobile executable project will automatically pick up correct dependency target.
And "net8.0" (or if you have a dedicated "net8.0-windows") desktop executable project will just fallback to compatible dependency target.

**Reply by maxkatz6** [reply]
Refresh rate shouldn't be as affected, native control render will be processed completely on its own.
